File: pagingProj/src/board/BoardDAO.java
/*
 *   insert 
 *   delete
 *   detail
 *   modify
 * 	 list
 */
package board;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import common.ConnectionPool;

public class BoardDAO {
	
	public int listBoardCount() throws Exception {
		Connection con = null;
		PreparedStatement pstmt = null;
		try {
			con = ConnectionPool.getConnection();
			StringBuffer sql = new StringBuffer();
			sql.append("select count(no) ")
			   .append("  from PAGING_BOARD ");
			pstmt = con.prepareStatement(sql.toString());
			ResultSet rs = pstmt.executeQuery();
			while (rs.next()) {
				// 첫번째 컬럼에 있는거 반환해라
				return rs.getInt(1);
			}
		} catch (Exception e) {
			throw e;
		} finally {
			try {
				pstmt.close();
			} catch (Exception e) {}
			ConnectionPool.releaseConnection(con);
		}
		return 0;
	}
	
	public List<BoardDomain> listBoard(Page page) throws Exception {
		List<BoardDomain> list = new ArrayList<>();
		Connection con = null;
		PreparedStatement pstmt = null;
		try {
			con = ConnectionPool.getConnection();
			StringBuffer sql = new StringBuffer();
			sql.append("select * ")
				.append("    from (select rownum runum, a.*")
				.append("     from (select * ")
				.append("   from PAGING_BOARD ")
				.append("        order by no desc)  a)")
				.append("where runum between ? and ?  ");
			
			
			pstmt = con.prepareStatement(sql.toString());
			int index = 1;
			pstmt.setInt(index++, page.getBegin());
			pstmt.setInt(index++, page.getEnd());
			ResultSet rs = pstmt.executeQuery();
			while (rs.next()) {
				BoardDomain board = new BoardDomain();
				board.setNo(rs.getInt("no"));
				board.setTitle(rs.getString("title"));
				board.setWriter(rs.getString("writer"));
				board.setContent(rs.getString("content"));
				board.setRegDate(rs.getTimestamp("reg_date"));
				list.add(board);
			}
		} catch (Exception e) {
			throw e;
		} finally {
			try {
				pstmt.close();
			} catch (Exception e) {}
			ConnectionPool.releaseConnection(con);
		}
		return list;
	}

}

File: javascript/WebContent/10_node/09_express_exam/model/dao/boardDAO.js
var con = require('../db-connect');


var boardDAO = {

    list : function(cb){
        var sql = `select * from tb_board order by no desc`;
        con.query(sql, function (err, rows, fields) {
            if (err) {
                cb(err);
                return;
            }
            cb(null, rows); // err를 구분하기 위한 null
        });
    },

    detail : function(arr, cb){
        var sql = `select * from tb_board where no = ?`;
        con.query(sql, arr, function (err, row, fields) {
            if (err) {
                cb(err);
                return;
            }
            cb(null, row[0]);
        });
    },

    delete : function(){
        var sql = `delete from tb_board where no = ?`;
        con.query(sql, arr, function (err, row, fields) {
            if (err) {
                cb(err);
                return;
            }
            cb(null, row[0]); //리다이렉션
        });
    }

}



module.exports = boardDAO;